docs/design/sparse-v2.md

```diff
@@ -0,0 +1,145 @@
+determine whether or not the current Sparse Patterns set tracked in the View is
+stale or not. SparsePatterns can thus be modified with `--ignore-working-copy`
+to support incremental edits without splatting the results into the working
+copy (expensive!) on every operation.
+
+### Type Evolution
+
+Before all else, SparsePatterns should be moved into a proper struct with
+encapsulation, rather than a bare list of RepoPaths. This will allow further
+evolutions of the type without breaking legacy clients that only know how to
+provide a list of prefixes or receive a Matcher.
+
+Sparse Patterns in text form will graduate to a `<type>:<path>` syntax, where
+`<path>` will be interpreted as `include:<path>` for backwards-compatibility on
+the command-line. Any unfortunate souls with sparse patterns where the first
+path component includes a colon will be safe (the migration will preserve the
+paths), but will need to use explicit syntax going forward when running
+`jj sparse` commands. Inside JJ, `<type>` will be a strongly typed enum with a
+Proto equivalent.
+
+The relevant types in Sparse Patterns v2 will be:
+
+- `include:<path>`: Match a specific prefix/dir
+- `exclude:<path>`: Exclude a specific prefix/dir
+- `files:<path>`: Include only the files in `<path>`, not subdirectories
+
+The `jj sparse` family of commands will be altered to support these specific
+patterns for adding and removing. We may also consider adding special flags
+like `jj sparse set --exclude <path>` as a short-hand for
+`jj sparse set --add exclude:<path>`.
+
+#### Exclusion Ordering
+
+The ordering of include and exclude rules in a SparsePattern matters. For the
+purpose of evaluation, any rule whose path is a suffix of another rule's path
+is treated as evaluated _after_ the other rule. Thus, the set of rules:
+
+```
+include:foo
+exclude:foo/bar
+include:foo/bar/baz
+exclude:foo/bar/baz/qux
+```
+
+Is interpreted to match "foo/file.txt", not-match "foo/bar/file.txt", match
+"foo/bar/baz/file.txt", and non-match "foo/bar/baz/qux/file.txt". If a user
+specifies both `include:foo` and `exclude:foo`, the exclusion will override the
+inclusion for the purpose of determinism. Internally, this will likely be
+represented as a tree of nested rules keyed by path component, but the
+representation is immaterial to the definition.
+
+### Mapping
+
+For various reasons, it is desirable to be able to transparently rename a
+directory foo -> bar in the working copy without actually renaming the directory
+in the public repo. Sparse Patterns are the ideal place to implement such a
+feature, since they are deeply tied to the working copy.
+
+Sparse Patterns will gain a 'map' field which is functionally a list of
+`(<type>, <src_path>, <dst_path>)` tuples. `<type>` will be an enum indicating
+'recursive' or 'non-recursive', while `<src_path>` and `<dst_path>` will be
+RepoPaths indicating the source directory being renamed, and its renamed
```

Comment:
Could you add a incomplete proto here? I think I get what you want to say but someone who just encounters this file while browsing, will have a bunch of questions.

Reply:
I revised this with a bunch of Rust code blocks following the other doc, is this more of what you were looking for?

Reply:
Yes, definitely.

Reply:
I lost the 'non-recursive' bit in the rewrite, added that back.